File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return False
        q = deque([(root, -1001, 1001)])
        while q:
            for i in range(len(q)):
                node, mini, maxi = q.popleft()
                if not (mini < node.val < maxi):
                    return False
                if node.left:
                    q.append((node.left, mini, node.val))
                if node.right:
                    q.append((node.right, node.val, maxi))

        return True
        
        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        std::queue<TreeNode*> st;
        if (root)
            st.push(root);
        while (st.size()) {
            auto* tn{st.front()};
            st.pop();
            std::swap(tn->left, tn->right);
            if (tn->left) st.push(tn->left);
            if (tn->right) st.push(tn->right);
        }
        return root;
    }
};
